syntax = "proto3";

package frances;

enum CrimeSceneFeatureType {
    WALL = 0;
    CORNER = 1;
    WINDOW = 2;
    CHAIR = 3;
    BED = 4;
    CARPET = 5;
    PLANT = 6;
    TV = 7;
    TABLE = 8;
}

enum PositionType {
    OCCUPIABLE_SPACE = 0;
    BLOCKED_SPACE = 1;
    VERTICAL_BOUNDARY = 2;
    HORIZONTAL_BOUNDARY = 3;
}

enum Gender {
    UNSPECIFIED_GENDER = 0;
    FEMALE = 1;
    MALE = 2;
}

enum Role {
    UNSPECIFIED_ROLE = 0;
    SUSPECT = 1;
    VICTIM = 2;
    MURDERER = 3;
}

message IntArray {
    repeated int32 values = 1;
}

message Coordinate {
    int32 row = 1;
    int32 column = 2;
}

message Room {
    int32 id = 1;
    string name = 2;
}

message CrimeSceneFeature {
    CrimeSceneFeatureType type = 1;
    PositionType position_type = 2;
    repeated Coordinate coordinates = 3;
}

message CrimeScene {
    repeated Room rooms = 1;
    repeated IntArray floor_plan = 2;
    repeated CrimeSceneFeature features = 3;
}

message Person {
    int32 id = 1;
    string name = 2;
    Gender gender = 3;
    Role role = 4;
    Coordinate coordinate = 5;
}

message RoomClue {
    int32 room_id = 1;
    bool is_occupied = 2;
}

message PersonClue {
    int32 person_id = 1;
    bool negate = 2;
    oneof type {
        int32 room_id = 3;
        CrimeSceneFeatureType beside = 4;
        CrimeSceneFeatureType on = 5;
        CrimeSceneFeatureType same_row = 6;
        CrimeSceneFeatureType same_column = 7;
        CrimeSceneFeatureType same_room = 8;
    }
}

message Clue {
    oneof type {
        RoomClue room_clue = 1;
        PersonClue person_clue = 2;
    }
}

message Puzzle {
    string name = 1;
    CrimeScene crime_scene = 2;
    repeated Person people = 3;
    repeated Clue clues = 4;
}